### ISSUE TITLE
Database not opened on second query (SQLNonTransientConnectionException, PersistenceException)

### ISSUE THREAD

**Issue by bajicdusko** [requery/requery#470]
I have followed examples to set-up requery in my android app.

Dagger AppModule.
```
@Provides
    @Singleton
    public ReactiveEntityStore<Persistable> provideReactiveEntityStore(EurekaApplication eurekaApplication) {
        DatabaseSource source = new DatabaseSource(eurekaApplication, Models.DEFAULT, 1);
        Configuration configuration = source.getConfiguration();
        return ReactiveSupport.toReactiveStore(new EntityDataStore<Persistable>(configuration) {
        });
    }
```

First query is executed here (to check if user exists in the database). UserRequeryRepository interface is provided from another Dagger module with ReactiveEntityStore as dependency.
This method actually throws NoSuchElementException (which is some other, unrelated issue)
```
public UserRequeryRepositoryImpl(ReactiveEntityStore<Persistable> data) {
        this.data = data;
    }

    @Override
    public Flowable<User> getUser() {
        return Flowable.fromCallable(new Callable<UserDb>() {
            @Override
            public UserDb call() throws Exception {
                UserDb userDb = data.select(UserDb.class).get().firstOrNull();
                if (userDb != null) {
                    return userDb;
                }

                throw new UserNotFoundException();
            }
        })
```

Second query is executed after user successfully login:
```
@Override
    public Flowable<User> saveUser(User user, String password, boolean rememberMe) {
        return data.select(UserDb.class).get().maybe()
                .defaultIfEmpty(new UserDb())
                .map(userDb -> {
                    if (!userDb.getUsername().equals(user.getUsername())) {
                        clearDatabase();
                    }
                    return user;
                })
                .map(user1 -> {
                    UserDb userDb = new UserDb();
                    DataMapperFactory.createRequery(user1).applyToDb(user1, userDb);
                    return userDb;
                })
                .flatMapSingle(data::insert)
                .map(userDb -> (User) DataMapperFactory.createRequery(userDb).to(userDb))
                .toFlowable();
}
```

This query fails on first line `data.select(UserDb.class)`.
Here is the stack trace:
```
io.requery.PersistenceException: java.sql.SQLNonTransientConnectionException
E/LoginPresenterImpl:     at io.requery.sql.EntityDataStore.checkConnectionMetadata(EntityDataStore.java:576)
E/LoginPresenterImpl:     at io.requery.sql.EntityDataStore$DataContext.read(EntityDataStore.java:643)
E/LoginPresenterImpl:     at io.requery.sql.EntityDataStore.select(EntityDataStore.java:443)
E/LoginPresenterImpl:     at io.requery.reactivex.WrappedEntityStore.select(WrappedEntityStore.java:256)
E/LoginPresenterImpl:     at com.infinitetouch.eureka.data.repository.user.impl.UserRequeryRepositoryImpl.saveUser(UserRequeryRepositoryImpl.java:55)
E/LoginPresenterImpl:     at com.infinitetouch.eureka.data.repository.user.impl.LoginApiRepositoryImpl$2.apply(LoginApiRepositoryImpl.java:76)
E/LoginPresenterImpl:     at com.infinitetouch.eureka.data.repository.user.impl.LoginApiRepositoryImpl$2.apply(LoginApiRepositoryImpl.java:73)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableFlatMap$MergeSubscriber.onNext(FlowableFlatMap.java:131)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableOnBackpressureBuffer$BackpressureBufferSubscriber.drain(FlowableOnBackpressureBuffer.java:186)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableOnBackpressureBuffer$BackpressureBufferSubscriber.onNext(FlowableOnBackpressureBuffer.java:111)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.onNext(FlowableSubscribeOn.java:97)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:90)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableDoOnEach$DoOnEachSubscriber.onNext(FlowableDoOnEach.java:90)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableOnErrorNext$OnErrorNextSubscriber.onNext(FlowableOnErrorNext.java:68)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableOnBackpressureLatest$BackpressureLatestSubscriber.drain(FlowableOnBackpressureLatest.java:128)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableOnBackpressureLatest$BackpressureLatestSubscriber.onNext(FlowableOnBackpressureLatest.java:67)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableFromObservable$SubscriberObserver.onNext(FlowableFromObservable.java:54)
E/LoginPresenterImpl:     at retrofit2.adapter.rxjava2.BodyObservable$BodyObserver.onNext(BodyObservable.java:51)
E/LoginPresenterImpl:     at retrofit2.adapter.rxjava2.BodyObservable$BodyObserver.onNext(BodyObservable.java:37)
E/LoginPresenterImpl:     at retrofit2.adapter.rxjava2.CallExecuteObservable.subscribeActual(CallExecuteObservable.java:43)
E/LoginPresenterImpl:     at io.reactivex.Observable.subscribe(Observable.java:10700)
E/LoginPresenterImpl:     at retrofit2.adapter.rxjava2.BodyObservable.subscribeActual(BodyObservable.java:34)
E/LoginPresenterImpl:     at io.reactivex.Observable.subscribe(Observable.java:10700)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableFromObservable.subscribeActual(FlowableFromObservable.java:31)
E/LoginPresenterImpl:     at io.reactivex.Flowable.subscribe(Flowable.java:12901)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableOnBackpressureLatest.subscribeActual(FlowableOnBackpressureLatest.java:31)
E/LoginPresenterImpl:     at io.reactivex.Flowable.subscribe(Flowable.java:12901)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableOnErrorNext.subscribeActual(FlowableOnErrorNext.java:38)
E/LoginPresenterImpl:     at io.reactivex.Flowable.subscribe(Flowable.java:12901)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableDoOnEach.subscribeActual(FlowableDoOnEach.java:48)
E/LoginPresenterImpl:     at io.reactivex.Flowable.subscribe(Flowable.java:12901)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableDoOnEach.subscribeActual(FlowableDoOnEach.java:48)
E/LoginPresenterImpl:     at io.reactivex.Flowable.subscribe(Flowable.java:12901)
E/LoginPresenterImpl:     at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
E/LoginPresenterImpl:     at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:59)
E/LoginPresenterImpl:     at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:51)
E/LoginPresenterImpl:     at java.util.concurrent.FutureTask.run(FutureTask.java:237)
E/LoginPresenterImpl:     at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
E/LoginPresenterImpl:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
E/LoginPresenterImpl:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
E/LoginPresenterImpl:     at java.lang.Thread.run(Thread.java:761)
E/LoginPresenterImpl: Caused by: java.sql.SQLNonTransientConnectionException
E/LoginPresenterImpl:     at io.requery.android.sqlite.DatabaseSource.getConnection(DatabaseSource.java:173)
E/LoginPresenterImpl:     at io.requery.android.sqlite.DatabaseSource.getConnection(DatabaseSource.java:242)
E/LoginPresenterImpl:     at io.requery.sql.EntityDataStore$DataContext.getConnection(EntityDataStore.java:623)
E/LoginPresenterImpl:     at io.requery.sql.EntityDataStore.checkConnectionMetadata(EntityDataStore.java:562)
E/LoginPresenterImpl: 	
... 40 more
```

Problem is here of course, because `db.isOpen == false` and I don't know why.
DatabaseSource.class
```
private Connection getConnection(SQLiteDatabase db) throws SQLException {
        synchronized(this) {
            if(!db.isOpen()) {
                throw new SQLNonTransientConnectionException();
            } else {
                return new SqliteConnection(db);
            }
        }
    }
```
Thank you for your help.

**Comment by bajicdusko** [comment]
I will reply to this issue just in case that someone else find it helpful.
Cause of the problem above is in my entities. I have created BaseEntity and UserEntity.

```
public abstract class AbstractBaseEntity implements Persistable{
    @Key
    long id;
}

public abstract class AbstractUserDb extends BaseEntity{
   String name;
}
```

After I have removed AbstractBaseEntity and modified all other entities to implement Persistable, another issues showed up, related to @OneToMany and @ManyToOne relations. I guess that because of structure above, ReQuery was not able to detect all other problems. 

Closing this issue.
